### PR TITLE
Support json formatted strings and Hash/Array merge lookups

### PR DESCRIPTION
**Pull request by abooitt**

### CODE REVIEW

**Comment by jsok** [comment]
I like this idea, my initial comment would be that I think this behaviour should be configurable.
JSON data in vault may or may not be intended to be consumed by hiera.

**Comment by abooitt** [comment]
Thanks! I will add configurability.
And what do you think about the addition to use the :hierarchy: list on top of the vault mounts list?

**Comment by abooitt** [comment]
Made it optional. And edited README.md accordingly.
Notice that json parsing is currently only attempted on the :default_field values.
Looking forward to your comments.

File: README.md [hunk 1]
```diff
@@ -48,9 +48,30 @@ To do this, set:
...
 
-    vault write secret/foo value=bar other=baz
+    vault write secret/foo value=bar
+
+The hiera lookup for `foo` will return just "bar" as a string, but only if it is the only field.
+If `foo` contains more fields, a Hash will be returned, just like with the default behaviour.
+
+### json parsing of single values - optional
+Only applicable when :default_field is used.
+To use json parsing, set, for example:
```
Comment (jsok) [inline README.md]: You need an empty new line before the indented code snippet or it doesn't render correctly.

File: lib/hiera/backend/vault_backend.rb [hunk 2]
```diff
@@ -61,13 +82,21 @@ def lookup_generic(key, scope)
           return nil if secret.nil?
 
           Hiera.debug("[hiera-vault] Read secret: #{key}")
-          if @config[:default_field]
+          if @config[:default_field] and secret.data.has_key?(@config[:default_field].to_sym) and secret.data.length == 1
```
Comment (jsok) [inline lib/hiera/backend/vault_backend.rb]: Why is the `length ==`1 check necessary?

e.g. if `:default_field = "value"` and you have `vault write secret/key value=x other=y else=z`, then the `other` and `else` fields should just be ignored rather.

File: README.md [hunk 3]
```diff
@@ -81,6 +102,8 @@ into their own mount. This could be achieved with a configuration like:
                 - secret
 
 
+The :hierarchy: entries from the hiera configuration are used for each mount.
```
Comment (jsok) [inline README.md]: Put the `:hierarchy:` keyword in backticks please.

File: README.md [hunk 4]
```diff
@@ -48,9 +48,30 @@ To do this, set:
...
 
-The hiera lookup for `foo` will return just "bar" as a string.
+The hiera lookup for `foo` will return an array. When used in Array lookups (hiera_array), all occurences of `foo` will be merged into a single array.
+
+In case the single field does not contain a parseable json string, the string will be returned as is. When used in Hash lookups, this will result in an error as normal.
+
+
+### Lookup type behavior
+
+In case Array or Hash lookup is done, usual array or hash merging takes place based on the configured merge_behavior setting.
```
Comment (jsok) [inline README.md]: `merge_behavior` in backticks here as well

File: README.md [hunk 5]
```diff
@@ -48,9 +48,30 @@ To do this, set:
 
 For example:
 
-    vault write secret/foo value=bar other=baz
+    vault write secret/foo value=bar
+
+The hiera lookup for `foo` will return just "bar" as a string, but only if it is the only field.
+If `foo` contains more fields, a Hash will be returned, just like with the default behaviour.
+
+### json parsing of single values - optional
```
Comment (jsok) [inline README.md]: Since this option depends on `default_field`, can this be a sub-section instead, i.e. `####` rather than `###`.

Also, JSON in uppercase.

File: README.md (hunk 1 above)
Comment (abooitt) [inline README.md]: Fixed in next commit.

File: lib/hiera/backend/vault_backend.rb (hunk 2 above)
Comment (abooitt) [inline lib/hiera/backend/vault_backend.rb]: Ignoring data that is in vault does not make sense to me. Ignoring the additional fields makes it impossible to retrieve the contents of those fields with hiera when using a default_field. I'd leave it to the consumer to decide whether to ignore the additional fields.
But I'll make the non-ignoring behavior optional as well, to make sure default behavior stays backwards compatible.

File: README.md (hunk 3 above)
Comment (abooitt) [inline README.md]: Fixed in next commit.

File: README.md (hunk 4 above)
Comment (abooitt) [inline README.md]: Fixed

File: README.md (hunk 5 above)
Comment (abooitt) [inline README.md]: Fixed both.

File: lib/hiera/backend/vault_backend.rb (hunk 2 above)
Comment (jsok) [inline lib/hiera/backend/vault_backend.rb]: Point taken, the example I'm imaging is setup like `:default_field = "hiera_only_value"`.

**Comment by jsok** [comment]
OK, I'm happy with the new functionality.

The double negative of `:no_ignore_additional` is hard to follow.
What do you think about options/UI (which takes a bit of inspiration from the `merge_behavior` option of hiera itself):
- `:default_field` - remains the same
- `:default_field_parse` - either `string` (default) or `json`
- `:default_field_behaviour` - either `ignore` or `include` (default) -- basically does the same thing a `::no_ignore_additional`

**Comment by abooitt** [comment]
Refactored the new config variables. The compound condition is now somewhat more readable. :)
And now, default behavior is exactly the same as before: if the default field is not present, nil is returned.

**Comment by abooitt** [comment]
Thanks for your attention so far. I appreciate that.
Wondering what your thoughts are about the latest commit, have you seen that I also added the behavior to loop through the hierarchy sources on top of the mounts?

While working further on this, I realized that it might be overkill to use the hierarchy list for lookups in vault. A similar behavior can be configured using the `:mounts` setting within the `:vault` config. 

For example, we have a list of 16 items in our hierarchy, which means for each hiera call within the puppet code 16 connections to vault will be created. And while we have quite some hiera lookups in each catalog compile, this easily results in hitting the open files limit in the vault service and, after increasing the open files limit, in source port exhaustion, in case vault is running on the puppetmaster and listens on localhost only.

As such I will make traversing the global hierarchy sources optional, as well.

Additionally I have been working on creating specific `hiera_vault`, `hiera_vault_array`, and `hiera_vault_hash` functions with corresponding configuration options, to be able to not use the vault backend for normal hiera calls, but only when using the specific `hiera_vault*` functions.
And with these functions it is then also possible to tell them to look in vault only.
For a sneak preview on this you can take a look in the `Flagged` branch in my fork.

Looking forward to your response, while continuing on the stuff.

**Comment by jsok** [comment]
Thanks Arnoud, I'm away on holidays at the moment so I won't be able to get around to this until mid next week. Sorry for the delay!

**Comment by abooitt** [comment]
Thanks for letting me know, Jonathan.
Enjoy your holidays.

**Comment by abooitt** [comment]
Hi Jonathan, just wondering when you will have some time.

File: lib/hiera/backend/vault_backend.rb [hunk 6]
```diff
@@ -10,6 +10,21 @@ def initialize()
         @config = Config[:vault]
         @config[:mounts] ||= {}
         @config[:mounts][:generic] ||= ['secret']
+        @config[:use_hierarchy] ||= 'no'
```
Comment (jsok) [inline lib/hiera/backend/vault_backend.rb]: I think we should use `true`/`false` here rather than yes/no.

File: README.md [hunk 7]
```diff
@@ -52,6 +52,59 @@ For example:
...
+    :vault:
+        :default_field: value
+        :default_field_behavior: only
+
+Then, when `foo` contains more fields in addition to `value`, a Hash will be returned, just like with the default behaviour.
+And, in case `foo` does not contain the `value` field, a Hash with the actual fields will be returned, as if `:default_field`
+was not specified.
+
+#### JSON parsing of single values - optional
+Only applicable when :default_field is used.
```
Comment (jsok) [inline README.md]: s/:default_field/`:default_field`/

**Comment by jsok** [comment]
I'm struggling to follow the `use_hierarchy` option's purpose. Can you give a more practical example of when it would be useful?

Can you consider removing it from this PR and opening in another? I'd love to get the JSON support out for you asap.

**Comment by abooitt** [comment]
Thank for your comments! I'll handle them.

I will take out the `use_hierarchy` functionality and put it in a new PR.

**Comment by jsok** [comment]
LGTM, will merge now!
